Feature/#41 공연정보 제공

CodingApe9 opened the pull request:
## 📄 Summary
> #41 
- 공연 목록 제공, 공연 상세 정보 제공 기능

## 🙋🏻 More
> queryDsl를 도입한 첫 pr 이네요!! 
사용법 숙지해주시고 의견 남겨주세요.

CodingApe9 committed:
feat : 공연 목록을 검색할 조건 dto작성

- 장르: 특정 장르에 해당하는 공연만 검색할 수 있다.
- 기간: 일간, 주간, 월간 단위를 선택하여 검색할 수 있다.
- 개수: 개수를 정해 검색할 수 있다.

해당하는 예외와 핸들러도 작성

CodingApe9 committed:
feat : 공연 목록 제공 기능

- 장르, 개수, 기간의 조건으로 공연의 목록을 제공할 수 있다.
- 장르를 지정하지 않으면 장르를 구분하지 않고 검색한다.
- 개수와 기간은 무조건 전달해야 한다.

CodingApe9 committed:
feat : 공연 api cors 설정

CodingApe9 committed:
fix : 공연 목록 검색 개수 버그 수정

- 10 ~ 50개를 받을 수 있게 수정

CodingApe9 committed:
feat : 공연 목록 검색 레포지토리 기능

CodingApe9 committed:
refact : 공연 목록 검색 응답 Dto 도입

- 엔티티 형태로 리턴하지 않고 dto를 사용하게 변경

CodingApe9 committed:
test : 공연 목록 검색 테스트 코드 추가

CodingApe9 committed:
Merge branch 'master' into feature/#41-공연정보-제공

CodingApe9 committed:
feat : 검색 기준 일자를 선택하는 기능

- 공연 목록을 검색시 기준 일자를 선택할 수 있다.

CodingApe9 committed:
test : 범위를 받아서 쿼리를 생성하게 테스트 코드 수정

- DateUnit -> Range

CodingApe9 committed:
test : 공역 목록 검색에 대한 인자 검증 테스트 코드 작성

- 날짜 단위(기간), 장르, 개수에 대한 테스트 코드 작성

CodingApe9 committed:
test : 공연 검색 테스트 코드 추가

CodingApe9 committed:
fix : 공연 목록 검색 버그 수정

- 조인에 누락된 조건을 추가 -> onStage와 reservation 조인 시 user의 null 여부를 체크함

CodingApe9 committed:
feat : 공연 목록 제공 요구사항 추가에 따른 레포지토리 수정

- 응답 dto 변경을 위한 레포지토리 수정
- 공연 정보에 더해 장소 정보까지 추가로 제공하는 기능 추가 작성

CodingApe9 committed:
feat : 공연 목록 제공 요구사항 추가에 따른 컨트롤러, 서비스 수정

- 레포지토리 반환 타입 dto 변경으로 인한 수정
- 공연 정보, 공연 장소 정보까지 추가로 제공하는 기능

CodingApe9 committed:
test : 공연 목록 제공 요구사항 추가에 따른 테스트 수정

CodingApe9 committed:
feat : @Size가 던지는 예외를 핸들링하는 핸들러 추가

- 공연 uuid는 36자리로 구성됩니다. 36자리의 스트링이 아니면 예외를 반환

CodingApe9 committed:
feat : 상세 공연 정보 제공을 위한 인터셉터 설정 추가

- 공연 상세 정보 제공은 유저 서비스가 아님

CodingApe9 committed:
Merge branch 'master' into feature/#41-공연정보-제공

CodingApe9 committed:
refact : dto 이름 변경

- SimplePlaceDTO -> PlaceDTO

CodingApe9 committed:
feat : 공연 상세 정보 제공을 위한 dto 작성

CodingApe9 committed:
feat : 공연 상세정보 제공 기능 작성

- 검색하려는 공연의 uuid는 36자리이어야 함
- 공연 정보, 공연의 좌석 등급 정보, 출연 스타의 정보를 제공함

cottonrose1011 commented on src/main/java/tback/kicketingback/performance/controller/PerformanceController.java:
```diff
@@ -0,0 +1,53 @@
+		return ResponseEntity.ok().body(getPerformancesResponse);
+	}
+
+	@GetMapping("/performance/{uuid}")
+	public ResponseEntity<DetailPerformanceDTO> getPerformance(
+		@PathVariable("uuid") @Size(min = 36, max = 36, message = "유효하지 않은 공연 uuid") String performanceUUID) {
+		DetailPerformanceDTO performance = performanceService.getPerformance(performanceUUID);
+		
+		return ResponseEntity.ok(performance);
+	}
```
uuid만 알고 있다면 권한 없이 디테일한 정보를 알 수 있는건가요?!

CodingApe9 replied:
네 맞아요.
공연 상세 정보가 원래 공개된 정보입니다!
다만 pk가 uuid라 다른 공연의 정보를 한 번에 확인은 못할겁니다.
auto increment int를 pk로 잡고 있으면 id를 추측할 수 있는데 uuid가 pk라 불가능합니다.

pwrwpw commented on src/main/java/tback/kicketingback/performance/dto/DateUnit.java:
```diff
@@ -0,0 +1,44 @@
+		return new Range(firstDayOfMonth, lastDayOfMonth);
+	});
+
+	private final Function<LocalDate, Range> rangeCalculator;
```
함수형 인터페이스로 깔끔하게 잘 구현하셨네요

CodingApe9 replied:
함수형 최고! 🙌

pwrwpw commented on src/main/java/tback/kicketingback/performance/dto/GetPerformancesRequest.java:
```diff
@@ -0,0 +1,15 @@
+
+	@NotNull(message = "유효하지 않은 검색 개수: null [10 ~ 50의 양수] 필요")
+	@Positive(message = "유효하지 않은 검색 개수: [10 ~ 50의 양수] 필요")
+	Integer size
```
왜 두개 어노테이션을 쓰셨나요 둘중 하나만 달아도 될 것 같습니다

그리구 뭔가 예외 메시지랑 어노테이션이 맞지 않는 것 같습니다

CodingApe9 replied:
중복 검증되고 있었네요 @NotNull만 있어도 충분할거 같아요.
예외 메시지도 같이 수정했습니다. 👍

pwrwpw commented on src/main/java/tback/kicketingback/performance/controller/PerformanceController.java:
```diff
@@ -0,0 +1,53 @@
+
+	@GetMapping("/performance/{uuid}")
+	public ResponseEntity<DetailPerformanceDTO> getPerformance(
+		@PathVariable("uuid") @Size(min = 36, max = 36, message = "유효하지 않은 공연 uuid") String performanceUUID) {
```
혹시 이부분 String으로 굳이 받지 않고 UUID로 받아도 되지 않을까요??

CodingApe9 replied:
그렇겠네요 수정하는게 더 좋아보여요.
String 타입의 uuid를 사용하는 일도 없고 사이즈 검증만하고 바로 변환하니 차라리 UUID로 처음부터 받는게 더 좋을 것 같아요.
수정해서 올리겠습니다!

pwrwpw reviewed (commented):
고생하셨습니다 리뷰 한번 확인 해주세요

pwrwpw commented on src/main/java/tback/kicketingback/performance/dto/GetPerformancesRequest.java:
```diff
@@ -0,0 +1,15 @@
+import jakarta.validation.constraints.Positive;
+
+public record GetPerformancesRequest(
+	@NotNull(message = "유효하지 않은 검색 기간 [day|week|month] 필")
```
이부분도 어차피 도메인 로직에서 예외 메시지를 통해 던지는데 이 검증에선 null 체크 메시지만 보내면 될 것 같습니다!!

CodingApe9 replied:
아래 수정하면서 같이 수정했어요!! 😎

CodingApe9 committed:
refact : 공연 상세정보 제공 타입 변경

- String uuid -> UUID uuid

CodingApe9 committed:
refact : 사용하지 않는 예외 핸들러 삭제 및 새로운 예외 핸들러 추가

- @Size가 던지는 예외 ConstraintViolationException 삭제
- MethodArgumentTypeMismatchException 핸들러 작성

CodingApe9 committed:
refact : 코드 리팩토링

- 인자 중복 검증 삭제
- 검증 예외 메시지 수정

pwrwpw reviewed (approved):
수고하셨습니다
